sstables/sstable_set.cc

```diff
@@ -61,94 +61,422 @@ std::ostream& operator<<(std::ostream& os, const sstables::sstable_run& run) {
+    return _version->size();
+}
+
+void sstable_list::insert(shared_sstable sst) {
+    if (_version->get_next_versions().size()) {
+        _version = sstable_set_version_reference(_version);
+    }
+    _version->insert(sst);
+}
+
+void sstable_list::erase(shared_sstable sst) {
+    if (_version->get_next_versions().size()) {
+        _version = sstable_set_version_reference(_version);
+    }
+    _version->erase(sst);
+}
+
+bool sstable_list::contains(shared_sstable sst) const {
+    return _version->contains(sst);
+}
+
+bool sstable_list::empty() const {
+    return _version->size() == 0;
+}
+
+const sstable_set_version& sstable_list::version() const {
+    return *_version;
+}
+
+sstable_set::sstable_set(std::unique_ptr<sstable_set_impl> impl, schema_ptr s)
+    : _all(make_lw_shared<sstable_list>(std::move(impl), std::move(s))) {
+        // assumes impl contains no sstables
+}
+
+sstable_set::sstable_set(const sstable_set& x)
+    : _all(make_lw_shared<sstable_list>(*x._all)) {
+}
+
+sstable_set::sstable_set(sstable_set&& x) noexcept = default;
+
+sstable_set& sstable_set::operator=(const sstable_set& ssts) {
+    if (this != &ssts) {
+        *this = sstable_set(ssts);
+    }
+    return *this;
+}
+
+sstable_set& sstable_set::operator=(sstable_set&& ssts) noexcept {
+    if (this != &ssts) {
+        this->~sstable_set();
+        new (this) sstable_set(std::move(ssts));
+    }
+    return *this;
+}
+
+
+std::vector<shared_sstable> sstable_set::select(const dht::partition_range& range) const {
+    return _all->version().select(range);
+}
+
+// Return all runs which contain any of the input sstables.
+std::vector<sstable_run> sstable_set::select_sstable_runs(const std::vector<shared_sstable>& sstables) const {
+    return _all->version().select_sstable_runs(sstables);
+}
+
+lw_shared_ptr<const sstable_list> sstable_set::all() const {
+    return _all;
+}
+
+void sstable_set::insert(shared_sstable sst) {
+    _all->insert(sst);
+}
+
+void sstable_set::erase(shared_sstable sst) {
+    _all->erase(sst);
 }
 
 sstable_set::incremental_selector::~incremental_selector() = default;
 
 sstable_set::incremental_selector::incremental_selector(sstable_set::incremental_selector&&) noexcept = default;
 
-sstable_set::incremental_selector::selection
-sstable_set::incremental_selector::select(const dht::ring_position_view& pos) const {
+sstable_set::incremental_selector::incremental_selector(std::unique_ptr<incremental_selector_impl> impl, const schema& s, lw_shared_ptr<sstable_list> sstl)
+    : _impl(std::move(impl))
+    , _cmp(s)
+    , _sstl(std::move(sstl)) {
+}
+
+sstable_set::incremental_selector::selection sstable_set::incremental_selector::select(const dht::ring_position_view& pos) const {
     if (!_current_range_view || !_current_range_view->contains(pos, _cmp)) {
-        std::tie(_current_range, _current_sstables, _current_next_position) = _impl->select(pos);
+        std::vector<shared_sstable> _current_versioned_sstables;
```

Comment:
_ prefix is for data members, not locals.

Reply:
Ok